admin-js/src/App.js

```diff
@@ -77,6 +87,8 @@ if (STATE["js_module"]) {
     // The inline comment skips the webpack import() and allows us to use the native
     // browser's import() function. Needed to dynamically import a module.
     MODULE_LOADER = import(/* webpackIgnore: true */ STATE["js_module"]).then((mod) => {
+        for (const k of Object.keys(mod.g))
+            mod.g[k] = COMPONENTS[k]
```

Comment:
The main application, after dynamically importing the module, simply loops through the modules global object and inserts all the functions it asks for.

This means the module component will be using the actual same react functions as the main application, and therefore have the same context etc. required to make the component work correctly.

Reply:
@Dreamsorcerer coming from SO as discussed. Would like to have link or docs to maybe have an environment where I can reproduce your issue and try out things. If that's too much work you should not bother because there's a tiny chance I can really take a deep look at it. Will try tho. Thanks

Reply:
Thanks, the base project is in Python, if you're on Mac/Linux then you'll already have it installed, so should only take a couple of minutes to run.

- Clone this branch and `cd aiohttp-admin`.
- Install python dependencies: `pip install -r requirements.txt`
- Build the JS module: `cd admin-js; yarn install; yarn build`
- Try the example:
  - `cd ../examples`
  - `PYTHONPATH='..' python simple.py`
  - Go to http://localhost:8080/admin
  - You can login with admin/admin and see the basic project working.

If you click on one of the rows to get to the show view, you'll see the custom button created in the examples:
![image](https://github.com/aio-libs/aiohttp-admin/assets/2203121/7c2a73e5-02f6-47ae-b122-4fb3762e83bc)

The custom button is loaded from examples/custom-button.js. If you want to try building something with yarn or similar, you can just create a new project and copy the output file over custom-button.js.

Reply:
@Anthony-E-B I had an idea, maybe you might know if it's workable and how to go about it?

I was thinking that maybe I could find a way to replace react in a project's dependencies with some kind of shim.

e.g. I want to use a 3rd-party component, and they have react listed as a dependency. In my build configuration, I replace react with my custom library (no idea how this is done, but presumably it must be possible so people can use patched versions of a library etc.). That custom library has all the same functions as react, except that it will actually proxy the calls to the main application.

So, essentially we replace react in the build script in order to build the component without react being in the compiled module. Then when we import that module from the main application, the module will send all those react calls to the main application, which will use the version of react already compiled in there.

Reply:
Maybe this can be used for replacing react:
https://github.com/nerdchacha/module-replace-webpack-plugin

Although doesn't appear to be maintained. Then I just need to figure out a way to build the shim and proxy the calls through.

Reply:
@Anthony-E-B If you're interested, I've come up with a solution in #804.

I think the only thing that would be nice to improve, is having some way to import react-admin/react-router-dom components when they are not included in the shim (because adding all of react-admin to the shim increases the application size by 400KB) with some kind of fallback mechanism in the build process.